nomad/resource_job.go

```diff
@@ -258,109 +265,129 @@ func resourceJobRegister(d *schema.ResourceData, meta interface{}) error {
-			Pending:    []string{"monitoring_deployment", "monitoring_evaluation"},
-			Target:     []string{"job_scheduled_without_deployment", "deployment_successful"},
-			Refresh:    deploymentStateRefreshFunc(client, evalId),
-			Timeout:    timeout,
-			Delay:      10 * time.Second,
-			MinTimeout: 3 * time.Second,
-		}
-
-		state, err := stateConf.WaitForState()
+	if d.Get("detach") == false && resp.EvalID != "" {
+		log.Printf("[DEBUG] will monitor scheduling/deployment of job '%s'", *job.ID)
+		deployment, err := monitorDeployment(client, timeout, resp.EvalID)
 		if err != nil {
 			return fmt.Errorf(
-				"error waiting for job '%s' to deploy successfully: %s",
+				"error waiting for job '%s' to schedule/deploy successfully: %s",
 				*job.ID, err)
 		}
-
-		deployment := state.(*api.Deployment)
-
 		if deployment != nil {
 			d.Set("deployment_id", deployment.ID)
 			d.Set("deployment_status", deployment.Status)
 		} else {
-			d.Set("deployment_id", "")
-			d.Set("deployment_status", "")
+			d.Set("deployment_id", nil)
+			d.Set("deployment_status", nil)
 		}
 	}
 
 	return resourceJobRead(d, meta) // populate other computed attributes
 }
 
-// deploymentStateRefreshFunc returns a resource.StateRefreshFunc that is used to watch
-// the deployment from a job create/update
-func deploymentStateRefreshFunc(client *api.Client, initialEvalId string) resource.StateRefreshFunc {
+// monitorDeployment monitors the evalution(s) from a job create/update and,
+// if they result in a deployment, monitors that deployment until completion.
+func monitorDeployment(client *api.Client, timeout time.Duration, initialEvalId string) (*api.Deployment, error) {
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{MonitoringEvaluation},
+		Target:     []string{EvaluationComplete},
+		Refresh:    evaluationStateRefreshFunc(client, initialEvalId),
+		Timeout:    timeout,
+		Delay:      0,
+		MinTimeout: 3 * time.Second,
+	}
+
+	state, err := stateConf.WaitForState()
+	if err != nil {
+		return nil, fmt.Errorf("error waiting for evaluation: %s", err)
+	}
+
+	evaluation := state.(*api.Evaluation)
+	if evaluation.DeploymentID == "" {
+		log.Printf("[WARN] job has been scheduled, but there is no deployment to monitor")
+		return nil, nil
+	}
+
+	stateConf = &resource.StateChangeConf{
+		Pending:    []string{MonitoringDeployment},
+		Target:     []string{DeploymentSuccessful},
+		Refresh:    deploymentStateRefreshFunc(client, evaluation.DeploymentID),
+		Timeout:    timeout,
+		Delay:      0,
+		MinTimeout: 5 * time.Second,
+	}
+
+	state, err = stateConf.WaitForState()
+	if err != nil {
+		return nil, fmt.Errorf("error waiting for evaluation: %s", err)
+	}
+	return state.(*api.Deployment), nil
+}
+
+// evaluationStateRefreshFunc returns a resource.StateRefreshFunc that is used to watch
+// the evaluation(s) from a job create/update
+func evaluationStateRefreshFunc(client *api.Client, initialEvalId string) resource.StateRefreshFunc {
```

Comment:
Same here `initialEvalId` -> `initialEvalID`
```suggestion
func evaluationStateRefreshFunc(client *api.Client, initialEvalID string) resource.StateRefreshFunc {
```